src/osmet/unpacker.rs

```diff
@@ -0,0 +1,271 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::convert::{TryFrom, TryInto};
+use std::ffi::OsStr;
+use std::fs::{File, OpenOptions};
+use std::io::{self, ErrorKind, Read, Seek, SeekFrom, Write};
+use std::os::unix::ffi::{OsStrExt, OsStringExt};
+use std::path::{Path, PathBuf};
+use std::thread;
+
+use error_chain::bail;
+use openssl::hash::{Hasher, MessageDigest};
+use xz2::read::XzDecoder;
+
+use super::*;
+
+/// Path to OSTree repo of sysroot.
+const SYSROOT_OSTREE_REPO: &str = "/sysroot/ostree/repo";
+
+pub struct OsmetUnpacker {
+    thread_handle: Option<thread::JoinHandle<Result<()>>>,
+    reader: pipe::PipeReader,
+    length: u64,
+}
+
+impl OsmetUnpacker {
+    pub fn new(osmet: &Path, repo: &Path) -> Result<Self> {
+        let (_, osmet, packed_image) = osmet_file_read(&osmet)?;
+        Self::new_impl(osmet, packed_image, repo)
+    }
+
+    pub fn new_from_sysroot(osmet: &Path) -> Result<Self> {
+        let (_, osmet, punched_image) = osmet_file_read(&osmet)?;
+        Self::new_impl(osmet, punched_image, Path::new(SYSROOT_OSTREE_REPO))
+    }
+
+    fn new_impl(
+        osmet: Osmet,
+        packed_image: impl Read + Send + 'static,
+        repo: &Path,
+    ) -> Result<Self> {
+        let (reader, writer) = pipe::pipe();
+
+        let length = osmet.size;
+        let repo = repo.to_owned();
+        let thread_handle = Some(thread::spawn(move || -> Result<()> {
+            osmet_unpack_to_writer(osmet, packed_image, repo, writer)
+        }));
+
+        Ok(Self {
+            thread_handle,
+            reader,
+            length,
+        })
+    }
+
+    pub fn length(&self) -> u64 {
+        self.length
+    }
+}
+
+impl Read for OsmetUnpacker {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let n = self.reader.read(buf)?;
+        if n == 0 {
+            match self
+                .thread_handle
+                .take()
+                .expect("pending thread")
+                .join()
+                .expect("joining thread")
+            {
+                Ok(_) => Ok(0),
+                Err(e) => Err(io::Error::new(
+                    ErrorKind::Other,
+                    format!("while unpacking: {}", e),
+                )),
+            }
+        } else {
+            Ok(n)
+        }
+    }
+}
+
+pub(super) fn get_unpacked_image_digest(
+    xzpacked_image: &mut File,
+    partitions: &[OsmetPartition],
+    root: &Mount,
+) -> Result<(Sha256Digest, u64)> {
+    let mut hasher = Hasher::new(MessageDigest::sha256()).chain_err(|| "creating SHA256 hasher")?;
+    let repo = {
+        let mut root_path = root.mountpoint().to_path_buf();
+        root_path.push("ostree/repo");
+        root_path
```

Comment:
Isn't this just `let repo = root.mountpoint().join("ostree/repo")`?

Reply:
Thanks, tweaked!